Test untested functions and traits of  `Felt`

haroune-mohammedi opened the pull request:
## Description
Issue: #823

We added three tests for `num_traits::One`:

- [x] `one_value` to test `One::one`
- [x] `is_one` to test `One::is_zero`
- [x] `one_multiplicative_identity` to test the multiplicative identity:

```
x * 1 = x       ∀ x
1 * x = x       ∀ x
```
- [x] `zero_additive_identity` to test the additive identity: of `num_traits::Zero`:
``` 
x + 0 = x       ∀ x
0 + x = x       ∀ x
```

Tests for `num_traits::Signed` implementation
   
- [x] `felt_is_never_negative`
- [x] `non_zero_felt_is_always_positive`
- [x]  `non_zero_felt_signum_is_always_one`
- [x] `signum_of_zero_is_zero`
- [x]  `sub_abs`
- [x]  `abs`

Tests for `Integer`:
- [x] `divides_doesnt_panic`
- [x] `gcd_doesnt_panic`
- [x] `is_even_always_true`
- [x] `is_odd_always_false`

Tests for `FromPrimitive` and `ToPrimitive`:
- [x] `from_u64_and_to_u64_primitive`
- [x] `from_i64_and_to_i64_primitive`

Tests for:
- [x] `modpow_in_range`
- [x] `sqrt_in_range`
- [x] `sqrt_is_inv_square`

Tests for `Add`,  `AddAssign`,  `Add<u32>`, `Sub<u32>`, `Sub<usize>`:

- [x] `add_u32_in_range`
- [x] `add_u32_is_inv_sub`
- [x] `sub_u32_in_range`
- [x] `sub_u32_is_inv_add`
- [x] `sub_uszie_in_range`
- [x] `sub_usize_is_inv_add`
- [x] `add_in_range`
- [x] `add_is_inv_sub`
- [x] `add_assign_in_range`

We tested the reference variant for 

- [x] `sub_assign_in_range`
- [x] `neg_in_range`
- [x] `pow_in_range`

As a result of writing tests, we discoverd bugs and fixed the implementation of `is_even`, `is_odd` and `divides`

## Checklist
- [x] Linked to Github Issue
- [x] Unit tests added
- [ ] Integration tests added.
- [ ] This change requires new documentation.
  - [ ] Documentation has been added/updated.

fmoletta reviewed (approved):
LGTM

Jrigada reviewed (approved):
Thanks for the PR, LGTM

haroune-mohammedi commented:
I have some suggestions that involves changes to the tests I haven't wrote so I figured I should ask you guys first

- Use `Felt::max_value` instead of parsing `PRIME_STR` inside all the functions
- use ref implementation to implement owned ones instead of implementing the same thing twice
- Why are we only testing for in_range, why don't we test the result values of the operations like: add, neg, sub ..., 
- Why do we need to implement the same operation for `Felt`, `&Felt`, `u32`, `usize` ... check https://docs.rs/impl_ops/0.1.1/impl_ops/index.html that could be used to get rid of the boilerplate

haroune-mohammedi commented:
I've added tests for `Felt` and updated the description

@fmoletta @Jrigada would you please re-review and approve the workflows, also is there any way I can run the workflows without waiting approval in my fork ?

Oppen commented on felt/src/lib.rs:
```diff
@@ -1052,6 +1088,204 @@ mod test {
             let x = Felt::parse_bytes(x.as_bytes(), 10).unwrap();
             let y = Felt::parse_bytes(y.as_bytes(), 10).unwrap();
             prop_assert!(x.is_multiple_of(&y));
+            // prop_assert!(x.divides(&y));
```
Let's remove this.

Oppen commented on felt/src/lib.rs:
```diff
@@ -1052,6 +1088,204 @@ mod test {
+        fn divides_doesnt_panic(ref x in "(0|[1-9][0-9]*)", ref y in "(0|[1-9][0-9]*)") {
+            let x = Felt::parse_bytes(x.as_bytes(), 10).unwrap();
+            let y = Felt::parse_bytes(y.as_bytes(), 10).unwrap();
+            prop_assert!(x.divides(&y));
```
This test will fail for `y == 0`. I'm not 100% what should happen for `x == 0` if this is not returning a `Result<_>`, but we're passing it in the test as well.

haroune-mohammedi replied:
It won't fail, `is_multiple_of` and `divides` always return `true`, see [`bigint_felt.rs#L615`](https://github.com/Quadratic-Labs/cairo-rs/blob/ef74e80fe4f8f1156233418e3a2c381db5b14047/felt/src/bigint_felt.rs#L615)

Oppen replied:
That is odd. `0` shouldn't divide anything.

Oppen commented on felt/src/lib.rs:
```diff
@@ -1052,6 +1088,204 @@ mod test {
+        #[test]
+        fn is_even(ref x in "(0|[1-9][0-9]*)") {
+            let x = Felt::parse_bytes(x.as_bytes(), 10).unwrap();
+            prop_assert_eq!(x.is_even(), x.to_biguint().is_even());
```
This will fail for a bunch of even uints. Take `CAIRO_PRIME + 1` for example. That's even before taking the modulus, but it becomes `1` afterwards.

Oppen replied:
Sorry, I thought you were comparing with the `BigUint` created from the string, not from the `Felt` itself.

Oppen commented on felt/src/lib.rs:
```diff
@@ -1052,6 +1088,204 @@ mod test {
+        #[test]
+        fn is_odd(ref x in "(0|[1-9][0-9]*)") {
+            let x = Felt::parse_bytes(x.as_bytes(), 10).unwrap();
+            prop_assert_eq!(x.is_odd(), x.to_biguint().is_odd());
```
The same happens here, `CAIRO_PRIME + 2` is odd, but `CAIRO_PRIME + 2 mod CAIRO_PRIME` is `2` which is even.

haroune-mohammedi replied:
Same answer as `is_even`

Oppen commented on felt/src/lib.rs:
```diff
@@ -1052,6 +1088,204 @@ mod test {
+            let one = Felt::one();
+            prop_assert_eq!(&x, &(&x * &one));
+            prop_assert_eq!(&x, &(&one * &x));
+        }
```
Nit: add a newline.

Oppen commented on felt/src/lib.rs:
```diff
@@ -1052,6 +1088,204 @@ mod test {
+            prop_assert_eq!(&x, &(&x * &one));
+            prop_assert_eq!(&x, &(&one * &x));
+        }
+        #[test]
+        fn non_zero_felt_is_always_positive(ref x in FELT_NON_ZERO_PATTERN) {
+            let x = Felt::parse_bytes(x.as_bytes(), 10).unwrap();
+            prop_assert!(x.is_positive())
+        }
+
+        #[test]
+        fn felt_is_never_negative(ref x in FELT_PATTERN) {
+            let x = Felt::parse_bytes(x.as_bytes(), 10).unwrap();
+            prop_assert!(!x.is_negative())
+        }
+
+
+        #[test]
+        fn non_zero_felt_signum_is_always_one(ref x in FELT_NON_ZERO_PATTERN) {
+            let x = Felt::parse_bytes(x.as_bytes(), 10).unwrap();
+            let one = Felt::one();
+            prop_assert_eq!(x.signum(), one)
+        }
```
IIRC, the `is_positive`, `signum` and `is_negative` for type `Felt` are meant for the signed representation, which reads a field element as the range `(-CAIRO_PRIME/2, CAIRO_PRIME/2)`, where `x > CAIRO_PRIME/2` represents a negative field element. This would make these assumptions wrong.

Oppen replied:
There was the intent to make a separate type for singed field elements to avoid this confusion, I don't exactly remember if it was applied (I'll check later today and come back with a definitive answer).

Oppen replied:
Effectively it still works as I detailed in the first comment.

haroune-mohammedi replied:
what should be done about them ?

Oppen replied:
I'd say the practical thing is to split each test in two and use a `prop_assume!` to fix the range. One test checks for negatives, one test checks for positives. The `0` case can be made a normal unit test.

haroune-mohammedi replied:
But that's not what is implemented in `Signed`, do you mean we have to change the implementation and make felts where `x > CAIRO_PRIME/2` negative ? What about having a separate `SignedFelt` type, isn't it a better design than messing with the implementation of `Felt` ?

Oppen replied:
Yes, the idea was to eventually refactor into a `SignedFelt` type that you could build from `Felt`. But for now it is that way. It should be documented and tested as that until we do the refactor. We're aware as it is right now it's unintuitive, but we prioritized fixing quickly fast over having it be beautiful, with our sight on refactoring it afterwards.
If you'd rather have it that way __now__ so tests can be more obvious, we can either see if someone on our side has time to do the refactor now (I think everyone has a task currently, so this is possible but unlikely), or maybe we can create an issue and you can assign it to yourself if you wish. The task would be to track where we actually care about sign (it's a few hints IIRC) as well as create the base shim.

Oppen commented on felt/src/lib.rs:
```diff
@@ -1052,6 +1088,204 @@ mod test {
+            prop_assert_eq!(&x.abs_sub(&y), &expected_abs_sub)
+        }
+
+        #[test]
+        fn abs(ref x in FELT_PATTERN) {
+            let x = Felt::parse_bytes(x.as_bytes(), 10).unwrap();
+            prop_assert_eq!(&x, &x.abs())
+        }
```
For the same reason as above, I'm not sure if this is true.

Oppen commented on felt/src/lib.rs:
```diff
@@ -1052,6 +1088,204 @@ mod test {
+        #[test]
+        fn add_u32_is_inv_sub(ref x in FELT_PATTERN, y in any::<u32>()) {
+            let x = &Felt::parse_bytes(x.as_bytes(), 10).unwrap();
+            let expected_y = (x.clone() + y - x).to_u32().unwrap();
```
`expected_y` does not necessarily fit in a `uint32`. `Felt(CAIRO_PRIME - 1) - u32::MAX` is still greater than `u32::MAX`.

Oppen replied:
You should convert `y` to `BigUint` before the comparison instead.

haroune-mohammedi replied:
Why wouldn't it fit, we're subtracting `x` from `x`, we should be left with exactly `y` or am I missing something here ?

Oppen replied:
Ugh, didn't see the `- x` 🤦

Oppen commented on felt/src/lib.rs:
```diff
@@ -1052,6 +1088,204 @@ mod test {
+        #[test]
+        fn sub_u32_in_range(ref x in FELT_PATTERN, y in any::<u32>()) {
+            let x = Felt::parse_bytes(x.as_bytes(), 10).unwrap();
+            let p = BigUint::parse_bytes(PRIME_STR[2..].as_bytes(), 16).unwrap();
```
Don't we already have this prime parsed?

haroune-mohammedi replied:
it's being parsed in every function, I wanted to get it out or even use `Felt::max_value` instead but I didn't want to change before asking (I asked in the comments about it)

Oppen commented on felt/src/lib.rs:
```diff
@@ -1052,6 +1088,204 @@ mod test {
+        #[test]
+        fn sub_uszie_in_range(ref x in FELT_PATTERN, y in any::<usize>()) {
+            let x = Felt::parse_bytes(x.as_bytes(), 10).unwrap();
+            let p = BigUint::parse_bytes(PRIME_STR[2..].as_bytes(), 16).unwrap();
```
Ditto.

Oppen commented on felt/src/lib.rs:
```diff
@@ -1052,6 +1088,204 @@ mod test {
+        fn add_in_range(ref x in "(0|[1-9][0-9]*)", ref y in "(0|[1-9][0-9]*)") {
+            let x = &Felt::parse_bytes(x.as_bytes(), 10).unwrap();
+            let y = &Felt::parse_bytes(y.as_bytes(), 10).unwrap();
+            let p = &BigUint::parse_bytes(PRIME_STR[2..].as_bytes(), 16).unwrap();
```
Ditto.

Oppen commented on felt/src/lib.rs:
```diff
@@ -1052,6 +1088,204 @@ mod test {
+        fn add_assign_in_range(ref x in "(0|[1-9][0-9]*)", ref y in "(0|[1-9][0-9]*)") {
+            let mut x = Felt::parse_bytes(x.as_bytes(), 10).unwrap();
+            let y = Felt::parse_bytes(y.as_bytes(), 10).unwrap();
+            let p = &BigUint::parse_bytes(PRIME_STR[2..].as_bytes(), 16).unwrap();
```
Ditto.

Oppen reviewed (changes requested):
Hello @haroune-mohammedi, thanks for your contribution, it looks good.
However, there are a few changes I need you to make before we can merge it.
